bozohttpd: init

liclac opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

I would like to be able to run the bozohttpd web server on NixOS. It's popular on netbsd, where it's developed in-tree, and is designed to run via inetd.

Includes a couple of patches to make it build on non-netbsd platforms, which have been upstreamed but not yet included in a release.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

liclac committed:
bozohttpd: init

liclac commented:
Huh, this looks like a bmake bug on aarch64-linux:
https://github.com/NixOS/nixpkgs/pull/124210/checks?check_run_id=2651774181

r-rmcgibbo commented:
Result of `nixpkgs-review pr 124210` at caaee8a5 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>bozohttpd</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/servers/http/bozohttpd/default.nix:68:3:
```
   |
68 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

lukegb commented:
cc @alyssais @sternenseemann

I think sterni mentioned we have `fetchNetBSD` for the NetBSD stuff, does it make sense to expose it so it can be used here?

sternenseemann commented:
I suppose we'd need to generalize it to support `/pkgsrc` as well and of course expose it, but this seems all very possible and worthwhile.

<https://github.com/NixOS/nixpkgs/blob/db9dd99b6c11c0017e8de6859f44ce5a867161ce/pkgs/os-specific/bsd/netbsd/default.nix#L10-L15>

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+  # backport two unreleased commits to fix builds on non-netbsd platforms.
+  patches = [
+    # add missing `#include <stdint.h>`
+    # https://freshbsd.org/netbsd/src/commit/qMGNoXfgeieZBVRC
+    ./0001-include-stdint.h.patch
```
please fethpatch https://freshbsd.org/netbsd/src/commit/qMGNoXfgeieZBVRC/diff/raw

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+    # https://freshbsd.org/netbsd/src/commit/qMGNoXfgeieZBVRC
+    ./0001-include-stdint.h.patch
+
+    # BUFSIZ is not guaranteed to be large enough
+    # https://freshbsd.org/netbsd/src/commit/A4ueIHIp3JgjNVRC
+    ./0002-dont-use-host-BUFSIZ.patch
```
https://freshbsd.org/netbsd/src/commit/A4ueIHIp3JgjNVRC/diff/raw

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+  ];
+  patchFlags = [ "-p3" ];
+
+  buildInputs = [ openssl ] ++ optional (luaSupport) lua;
```
```suggestion
  buildInputs = [ openssl ]
    ++ optional luaSupport lua;
```

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+  buildInputs = [ openssl ] ++ optional (luaSupport) lua;
+  nativeBuildInputs = [ bmake groff ];
+
+  COPTS =
+    [
+      "-DNO_BLOCKLIST_SUPPORT" # unpackaged dependency: https://man.netbsd.org/blocklist.3
+    ]
+    ++ optional (!userSupport) "-DNO_USER_SUPPORT"
+    ++ optional (!dirIndexSupport) "-DNO_DIRINDEX_SUPPORT"
+    ++ optional (!dynamicContentSupport) "-DNO_DYNAMIC_CONTENT"
+    ++ optional (!luaSupport) "-DNO_LUA_SUPPORT"
+    ++ optional (!sslSupport) "-DNO_SSL_SUPPORT"
+    ++ optional (!cgiSupport) "-DNO_CGIBIN_SUPPORT"
+    ++ optional (htpasswdSupport) "-DDO_HTPASSWD";
```
```suggestion
  COPTS = [
    "-DNO_BLOCKLIST_SUPPORT" # unpackaged dependency: https://man.netbsd.org/blocklist.3
    ] ++ optional (!userSupport) "-DNO_USER_SUPPORT"
    ++ optional (!dirIndexSupport) "-DNO_DIRINDEX_SUPPORT"
    ++ optional (!dynamicContentSupport) "-DNO_DYNAMIC_CONTENT"
    ++ optional (!luaSupport) "-DNO_LUA_SUPPORT"
    ++ optional (!sslSupport) "-DNO_SSL_SUPPORT"
    ++ optional (!cgiSupport) "-DNO_CGIBIN_SUPPORT"
    ++ optional htpasswdSupport "-DDO_HTPASSWD";
```

liclac replied:
While I don't entirely disagree with this, `nixpkgs-fmt` does. Do you use a different formatting tool?

SuperSandro2000 replied:
nixpkgs-fmt does not enforce styling. It will happily accept both variants.

liclac replied:
Right, `nixpkgs-fmt` will accept both what I've done here:

```
VAR =
  [
    a
  ]
  ++ b
  ++ c;
```

and the annoyingly asymmetric:

```
VAR = [
  a
] ++ b
++ c;
```

but not what you suggest:

```
VAR = [
    a
  ] ++ b
  ++ c;
```

or what I'd personally prefer:


```
VAR = [
    a
  ]
  ++ b
  ++ c;
```

...hence my question about if you were using a different formatting tool which does accept this suggestion, or if you just typed it into the web UI by hand without checking.

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+    ++ optional (!cgiSupport) "-DNO_CGIBIN_SUPPORT"
+    ++ optional (htpasswdSupport) "-DDO_HTPASSWD";
+
+  _LDADD = [ "-lcrypt" "-lm" ]
+    ++ optional (luaSupport) "-llua"
+    ++ optionals (sslSupport) [ "-lssl" "-lcrypto" ];
```
```suggestion
  _LDADD = [ "-lcrypt" "-lm" ]
    ++ optional luaSupport "-llua"
    ++ optionals sslSupport [ "-lssl" "-lcrypto" ];
```

liclac replied:
While this would work here, eg. `!luaSupport` (a few lines above) is not valid without parentheses, and I'm not sure I like the idea of making this inconsistent. Is this a convention I'm not aware of?

SuperSandro2000 replied:
It is usually a good idea to not add any extra parentheses in ifs if they are not required but whatever.

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+
+  installPhase = ''
+    install -m755 -Dt $out/bin bozohttpd
+    install -m644 -Dt $out/share/man/man8 bozohttpd.8
```
Please use installManPage.

liclac replied:
This has been replaced entirely by bmake's own `make install` target.

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+    homepage = "http://www.eterna.com.au/bozohttpd/";
+    changelog = "http://www.eterna.com.au/bozohttpd/CHANGES";
+    license = licenses.bsd2;
+    maintainers = [ maintainers.embr ];
```
Please add yourself.

sternenseemann replied:
They did.

SuperSandro2000 replied:
I may suggest renaming the name in nixpkgs or adding embr as a display to GitHub because otherwise it is not really obvious.

alyssais replied:
People _really_ don't like it when you ask them to change their names (look how that went last time).  I think it's safe to assume in a new package with one maintainer listed, that that's the person submitting the PR.  In fact, you can tell very easily that that's the case, because OfBorg has added the "by: package mantainer" label here.

liclac replied:
It is my display name on github, and also my name.

liclac replied:
And when all else fails, you can always check the maintainer list: https://github.com/NixOS/nixpkgs/blob/f2f00af5b8e04a848c2b9ab8dcaf3e1c22a90dfb/maintainers/maintainer-list.nix#L2796-L2801

grahamc replied:
@SuperSandro2000 it is really not your responsibility to determine if the name people are using is okay, people can use whatever display names / maintainer names / github handles they want, and they don't need to match. Please don't push back on this in the future, and let them be.

SuperSandro2000 replied:
> I think it's safe to assume in a new package with one maintainer listed, that that's the person submitting the PR.

It already happened multiple times that newcomers copied files from another package and forgot to change the maintainer field. Just making sure that isn't happening.

> In fact, you can tell very easily that that's the case, because OfBorg has added the "by: package mantainer" label here.

Good reminder. Totally forgot about that.

> It is my display name on github, and also my name.

Maybe I got trolled by GitHubs UI. Don't know anymore.

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+, htpasswdSupport ? true
+}:
+
+let inherit (lib) optional optionals;
+in
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
```

sternenseemann replied:
This change would break evaluation…

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+  # http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/www/bozohttpd/distinfo
+  src = fetchurl {
+    url = "http://www.eterna.com.au/${pname}/${pname}-${version}.tar.bz2";
+    sha512 = "b838498626ffb7f7e84f31611e0e99aaa3af64bd9376e1a13ec16313c182eebfd9ea2c2d03904497239af723bf34a3d2202dac1f2d3e55f9fd076f6d45ccfa33";
```
Please use a base64 encoded sha256.

sternenseemann replied:
> canonical hashes can be found at: http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/www/bozohttpd/distinfo

There is value in using upstream checksums as is.

SuperSandro2000 commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+, inetutils
+, wget
+, openssl
+, userSupport ? true
+, cgiSupport ? true
+, dirIndexSupport ? true
+, dynamicContentSupport ? true
+, sslSupport ? true
+, luaSupport ? true
+, lua
+, htpasswdSupport ? true
```
Do we need all those options even? Can we just enable them by default?

```suggestion
, lua
```

liclac replied:
We could, and I'm enabling them all by default, but I figured if the upstream developer exposes and supports these knobs, it's easy enough for us to do so as well.

SuperSandro2000 replied:
We usually do not do this and just enable all common defaults people usually want to use.

liclac replied:
I feel like if these flags are there upstream, there's probably someone who uses them, especially as most of them are "opt-outs" rather than "opt-ins" - eg. lua support is enabled by default, but you can explicitly disable it. Especially in a web server, I can see why you'd would want to disable functionality that you're not using, and reduce its attack surface.

Is there a reason why you'd like to see them removed?

SuperSandro2000 replied:
> Is there a reason why you'd like to see them removed?

They add extra complexity and most of the time they are not tested while updating which causes them to bitrot and break over time especially if you use a complex combination of them.

niacat replied:
We test all of these upstream because we have a "small" target that builds an extremely small bozohttpd for extremely low memory machines. The only one that affects anything other than which fragmwnts of code are included is the "lua" and "ssl" options, which change dependencies. Since bozohttpd has to build entirely from the NetBSD base system, it is extremely unlikely that any new requirements would be added. I don't see how supporting them introduces any complexity.

sternenseemann replied:
What about a `bozohttpdMinimal` top level attribute with all optional features disabled by default? This way the “testing this still builds” concern should be alleviated a bit.

liclac replied:
Ensuring that packages do not bitrot when updated is firmly the maintainer's responsibility. If you're concerned for my workload, it's appreciated, but not necessary - I wouldn't have added these options if I wasn't confident that they would be maintainable. As nia says, they are unlikely to change.

But they could indeed use a CI pass, so I've added another `minimal` option and a `bozohttpd-minimal` top-level package, to roughly mirror the netbsd `small` target.

liclac replied:
@SuperSandro2000 Should I take your silence as that this has adequately addressed your concerns?

alyssais commented:
> I think sterni mentioned we have `fetchNetBSD` for the NetBSD stuff, does it make sense to expose it so it can be used here?

I don't think so, since there are release tarballs.

(I'd actually like fetchNetBSD to go away completely in favour of the
NetBSD release tarballs, but I didn't get to that yet.)

alyssais commented:
@supersandro2000 We can't generally use `fetchpatch` with CVS-based projects (including in this case -- I checked), because of the revision headers, which are incremented every time the file is changed, which will always cause a conflict if you apply patches out of order.

We could make `fetchpatch` support this, if we made it expose `filterdiff`'s `-#`/`--hunks` option, so we could just remove the first hunk when fetching CVS patches, but nobody has done that yet.

alyssais commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+
+  doCheck = true;
+  checkInputs = [ inetutils wget ];
+  checkFlags = [ "-dx" "VERBOSE=yes" ] ++ optional (!cgiSupport) "CGITESTS=";
```
`-dx ` and `VERBOSE=yes` can probably be removed, right?

liclac replied:
Done!

alyssais commented on pkgs/servers/http/bozohttpd/default.nix:
```diff
@@ -0,0 +1,91 @@
+  checkInputs = [ inetutils wget ];
+  checkFlags = [ "-dx" "VERBOSE=yes" ] ++ optional (!cgiSupport) "CGITESTS=";
+
+  installPhase = ''
```
If we add `"prefix=$(out)" "BINOWN="` to tell bmake about the output path, and that it shouldn't try to make the files owned by root, the normal `installPhase` will work, so you won't need to maintain a custom one. :)

liclac replied:
Done!

liclac commented:
Yeah, I actually used `fetchpatch` at first, but I couldn't find a way to not make the CVS headers clash other than by pulling down the patches and editing out those hunks.

It works, but it's a bit dirty, and I'd love to not do this, if anyone could tell me how.

liclac committed:
bozohttpd: use bmake flags instead of a custom installPhase

liclac committed:
bozohttpd: remove unnecessary checkFlags

liclac commented:
Removed the custom installPhase in favour of passing a few bmake flags. I think these should perhaps even live in the bmake hook itself - they should all Just Do The Right Thing in our context.

SuperSandro2000 commented:
> We can't generally use fetchpatch with CVS-based

Oh, I thought this is git. Sometimes I forget that there are other version control systems that are actually used somewhere.

liclac committed:
bozohttpd: resolve warnings, serve >2GB files on 32bit host

-D_DEFAULT_SOURCE *and* -D_GNU_SOURCE together resolve all warnings
about implicitly defined functions.

-D_LARGEFILE_SOURCE and -D_FILE_OFFSET_BITS=64 enable gcc's Large File
Extensions to allow a 32-bit host to serve files larger than 2GB.

Based on what Makefile.boot does, for platforms without bmake:
http://cvsweb.netbsd.org/bsdweb.cgi/src/libexec/httpd/Makefile.boot

liclac committed:
bozohttpd: add bozohttpd-minimal target

niacat commented:
> I suspect some of the `-D` flags we pass are also unnecessary for non-glibc, but they don't hurt either probably.

Are you sure about this patch? libm and libcrypt aren't glibc-exclusive, they're also needed on BSD systems. It would be better to have a portable test somewhere. Or just limit it to non-Darwin for now.

sternenseemann commented:
> Are you sure about this patch? libm and libcrypt aren't glibc-exclusive, they're also needed on BSD systems. It would be better to have a portable test somewhere. Or just limit it to non-Darwin for now. @sternenseemann

Not entirely. I know at least that `-lm` is unnecessary for musl and assumed it was the same for `-lcrypt`. I guess it'd be safer to just disable the `-lcrypt` flag for `libc == "libSystem"`.

liclac committed:
bozohttpd: minor formatting

The Nix style guide isn't exactly comprehensive, but this at least seems
slightly closer to the examples:

https://nixos.org/manual/nixpkgs/stable/#sec-syntax

I'm annoyed that nixpkgs-fmt as of writing won't accept:

  COPTS = [
    # ...
  ] ++ optional (...) "..."
    ++ optional (...) "...";

...which seems to be the actual convention, looking at eg. pkgs.nginx.

liclac committed:
bozohttpd: experimental darwin compat patch

As far as I can tell, libSystem on darwin provides a libm compatibility
symlink, but not a libcrypt one.

Prior art (eg. CPython) appears to use `stdenv.isDarwin` as a proxy for
`stdenv.hostPlatform.libc == "libSystem"`, but the latter also works in
cases where we're building on Darwin, but with eg. musl for some reason.

liclac commented:
I pushed an experimental libSystem compatibility patch, let's see if this works. :eyes:

I don't have a working Darwin machine handy right now, but as I remember it, libSystem makes both libm and libcrypt redundant, but only provides a compatibility symlink for the former.

sternenseemann commented:
Build on darwin works.

liclac commented:
@SuperSandro2000 hasn't replied for a week, even with my ping two days ago. According to the stats so prominently on display on their profile, they seem to have been plenty active on other PRs in this repo in that time.

If they're no longer interested in reviewing this PR, there a way we could move this along with @sternenseemann's review, or get someone else to take a look, rather than letting it go stale?

lukegb commented:
@ofborg build bozohttpd bozohttpd-minimal
